packages/gatsby-plugin-fastify/src/plugins/headers.ts

```diff
@@ -0,0 +1,44 @@
+import fp from "fastify-plugin";
+import { isMatch } from "picomatch";
+import type { FastifyPluginAsync, FastifyReply } from "fastify";
+import { getConfig } from "../utils/config";
+import { NEVER_CACHE_HEADER } from "../utils/constants";
+
+export const handleHeaders: FastifyPluginAsync = fp(async (fastify) => {
+  fastify.addHook("onSend", async function (request, reply: FastifyReply) {
+    // add all necessary headers
+    const {
+      server: {
+        headers,
+        features: {
+          headers: { useDefaultCaching },
+        },
+      },
+    } = getConfig();
+
+    // get the url without query params
+    const url = request.raw.url?.split(`?`)[0];
+
+    // handle directory URLs which resolve to index.html
+    if (
+      useDefaultCaching &&
+      (url?.endsWith("/") || // directories with trailing slash
+        (!url?.includes(".") && !url?.endsWith("/"))) // no trailing slash or file extension, so it's a directory
+    ) {
+      reply.headers(NEVER_CACHE_HEADER);
+    }
+
+    // handle pattern match caching
+    for (const [pattern, headerObject] of Object.entries(headers)) {
+      if (url && isMatch(url, pattern)) {
+        // set configured headers
+        reply.headers(headerObject);
+
+        // overwrite configured headers with SSR values
+        if (reply.pageQueryData?.serverDataHeaders) {
+          reply.headers(reply.pageQueryData.serverDataHeaders);
+        }
+      }
+    }
+  });
```

Comment:
I definitely see how this might be a problem though I personally like Option 3 😉😏

Joking aside, if option 3 isn't feasible, I would lean toward build time as well. I'll see about running the `benchmark.js` file tomorrow morning and compare (unless you have time during review tonight).

Reply:
Let's start with bench marks (I won't have time) and go from there. That script runs on the current branch. So recomendation would be to run on main a few times then checout this branch and run again. The more things you can close on your computer the more accurate the results will be. Basically close everything but a raw terminal and execute from there.

Reply:
Let's start with bench marks (I won't have time) and go from there. That script runs on the current branch. So recomendation would be to run on main a few times then checout this branch and run again. The more things you can close on your computer the more accurate the results will be. Basically close everything but a raw terminal and execute from there.